### DISCUSSION TITLE
Using ESP32-S3 as a USB Mass Storage Device

### DISCUSSION THREAD

**Discussion by Nico-GYW** [micropython/micropython#11282] [ESP32]
Hello,

I am currently working on a project using an ESP32-S3 and MicroPython. My goal is to use the ESP32-S3 as a USB Mass Storage Device (MSD) so that I can easily transfer files between my computer and the ESP32-S3.

I would like to be able to connect the ESP32-S3 to my computer via USB and have it appear as a removable drive, just like a regular USB flash drive. My ESP32-S3 is also connected to a MicroSD card using the built-in SD/MMC hardware. I would like this SD card to be accessible from my computer as the storage medium when using the ESP32-S3 as a USB MSD.

I have already connected the USB D+ and D- pins to a USB connector, and I am able to flash the ESP32-S3 via the USB pins and transfer files using Mpremote. 

However, I am not sure about the next steps to achieve the desired USB MSD functionality. I have looked into the TinyUSB library and CircuitPython, but it seems like TinyUSB is used by the MicroPython firmware itself and not directly accessible from MicroPython libraries. As for CircuitPython, I would prefer to stick with MicroPython for my project, as I find it more suitable for my needs.

My questions are as follows:

Is there a way to implement the USB MSD functionality using MicroPython on the ESP32-S3 without having to switch to CircuitPython?

Are there any existing MicroPython libraries or resources that can help me achieve this goal?

If not, could you provide some guidance on how to approach implementing this functionality with MicroPython?

Thank you in advance for your help!

**Comment by peterhinch** [comment]
See https://github.com/micropython/micropython/issues/8426 for a discussion of this issue.
